bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/protocol/RemoteControllerLegacy.java

```diff
@@ -190,100 +193,83 @@ private void readInitialInfo(Reader reader) throws RemoteControllerException {
-        }
+    public boolean closeApp() {
+        return false;
+    }
 
-        logger.debug("Command successfully sent");
+    public void getAppStatus(String id) {
     }
 
-    /**
-     * Send sequence of key codes to Samsung TV.
-     *
-     * @param keys List of key codes to send.
-     * @throws RemoteControllerException
-     */
-    @Override
-    public void sendKeys(List<KeyCode> keys) throws RemoteControllerException {
-        sendKeys(keys, 300);
+    public boolean noApps() {
+        return false;
+    }
+
+    private void logResult(String msg, Throwable cause) {
+        if (logger.isTraceEnabled()) {
+            logger.trace("{}: {}: ", host, msg, cause);
+        } else {
+            logger.debug("{}: {}: {}", host, msg, cause.getMessage());
+        }
     }
 
     /**
-     * Send sequence of key codes to Samsung TV.
+     * Send key code to Samsung TV.
      *
-     * @param keys List of key codes to send.
-     * @param sleepInMs Sleep between key code sending in milliseconds.
-     * @throws RemoteControllerException
+     * @param key Key code to send.
      */
-    public void sendKeys(List<KeyCode> keys, int sleepInMs) throws RemoteControllerException {
-        logger.debug("Try to send sequence of commands: {}", keys);
-
-        if (!isConnected()) {
-            openConnection();
+    public void sendKey(Object key) {
+        if (!(key instanceof KeyCode)) {
+            logger.warn("{}: Remote control legacy: unsupported command: {}", host, key);
+            return;
         }
-
-        for (int i = 0; i < keys.size(); i++) {
-            KeyCode key = keys.get(i);
+        logger.trace("{}: Try to send command: {}", host, key);
+        for (int i = 0; i < 2; i++) {
             try {
-                sendKeyData(key);
-            } catch (RemoteControllerException e) {
-                logger.debug("Couldn't send command", e);
-                logger.debug("Retry one time...");
-
-                closeConnection();
                 openConnection();
-
-                sendKeyData(key);
-            }
-
-            if ((keys.size() - 1) != i) {
-                // Sleep a while between commands
-                try {
-                    Thread.sleep(sleepInMs);
-                } catch (InterruptedException e) {
+                if (sendKeyData((KeyCode) key)) {
+                    logger.trace("{}: Command successfully sent", host);
                     return;
                 }
+            } catch (RemoteControllerException e) {
+                logResult("Couldn't send command", e);
             }
+            closeConnection();
+            logger.debug("{}: Retry send command {} attempt {}...", host, key, i);
         }
+        logger.warn("{}: Command Retrys failed", host);
+    }
 
-        logger.debug("Command(s) successfully sent");
+    public void sendKeyPress(KeyCode key, int duration) {
+        sendKey(key);
     }
 
-    @Override
+    @SuppressWarnings("null")
```

Comment:
Please fix in code, prevent suppressions

Reply:
Done